### ISSUE TITLE
Support search in LegacyRepository

### ISSUE THREAD

**Issue by taranlu-houzz** [python-poetry/poetry#2446]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: macOS 10.14.6
- **Poetry version**: Poetry version 1.0.5
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: https://gist.github.com/taranlu-houzz/d85b1ff99c8a0610ae7ed6911263038d

## Issue
When I use `poetry search <package from private repo>` it does not show up in the search results. However, using `poetry add <package from private repo>` works just fine.

## Config Notes
- I have a local pypi-server running using the package `pypiserver`.
- The package I am trying to search for is several levels down in a namespace package.
- This project is also a package that lives several levels down in a namespace package.
- I am able to publish to this local repo just fine using `poetry publish` with the appropriate flags.

**Comment by jduprey** [comment]
I have seen this too... it was a while ago, but I think my colleague dug into poetry and found that they pypi URL was hardcoded for search in poetry.. maybe.. that is from memory so I could be wrong.

Would love to see search fixed.  Poetry add/install seems to work fine with private repos now.

**Comment by WYishai** [comment]
@jduprey, it looks like you're right.
The relevant code is here:
https://github.com/python-poetry/poetry/blob/master/src/poetry/console/commands/search.py#L16
the `search` command creates a `PyPiRepository` with its default values, which cause it to use the global PyPi:
https://github.com/python-poetry/poetry/blob/master/src/poetry/repositories/pypi_repository.py#L52

Maybe a good-enough solution is to add an optional flag that specifies the remote repository URL. It should be very easy and simple to write (just a ~5 lines of code in the `search` command), and I think that in the cases you are using the `search` command you're probably not in the context of a specific repository...

**Comment by taranlu-houzz** [comment]
@WYishai While that would work, it seems like a bit of a half-solution to me. In order to work with the private repo, I have to explicitly add that as a source in the `pyproject.toml`. To me, that indicates that the added source should always be considered, since that is what happens when using `poetry add`.

**Comment by fbrodrigorezino** [comment]
@taranlu-houzz I didn't understand, you can search in private repos if you config you `poetry config`  with the private repository? 

For me it's not working, I have the repository on `poetry config --list` and when I use search it does not search in private repositories.

**Comment by neersighted** [comment]
This is currently by design -- `search` used to be implanted using the old XML-RPC API before Poetry started using PyPI's JSON API to roll our own. Whether or not this is feasible to implement with `LegacyRepository` (aka PEP 503 indexes) is unknown, but this is a feature request for that.

**Comment by lumen-sreejithr** [comment]
Hi @neersighted, @abn, Is there a solution or interim workaround to get this done till this PR gets merged?

**Comment by abn** [comment]
@lumen-sreejithr Can you please try the fix at #9949.

### Using pipx
```sh
pipx install --suffix=@9949 'poetry @ git+https://github.com/python-poetry/poetry.git@refs/pull/9949/head'
```

### Using a container (podman | docker)
```sh
podman run --rm -i --entrypoint bash python:3.13 <<EOF
set -xe
python -m pip install --disable-pip-version-check -q git+https://github.com/python-poetry/poetry.git@refs/pull/9949/head
poetry new foobar
pushd foobar
poetry source add pytorch https://download.pytorch.org/whl/cu113
poetry search pytorch
EOF
```

Please note that PyPI no longer supports search.

**Comment by lumen-sreejithr** [comment]
Hi @abn, Thanks for this. I've tested the version of poetry mentioned above and Please see the observations:

## Platform: Linux (Ubuntu)

Works perfectly with pytorch repo and locally hosted [pypi-server](https://pypi.org/project/pypiserver/)

```bash
(poetry-test) user@ubuntu:~/foobar$ poetry source add pytorch https://download.pytorch.org/whl/cu113
Adding source with name pytorch.
(poetry-test) user@ubuntu:~/foobar$ poetry search pytorch

pytorch-triton (2.3.0)

pytorch-triton (3.0.0)

pytorch-triton-rocm (2.0.0)

pytorch-triton-rocm (2.0.1)

pytorch-triton-rocm (2.1.0)

pytorch-triton-rocm (2.2.0)

pytorch-triton-rocm (2.3.0)

pytorch-triton-rocm (2.3.1)

pytorch-triton-rocm (3.0.0)

pytorch-triton-rocm (3.1.0)
```

**Local PyPI Server**

```bash
(poetry-test) user@ubuntu:~/foobar$ poetry source add pypi-dev http://localhost:8080/simple
Adding source with name pypi-dev.
(poetry-test) user@ubuntu:~/foobar$ poetry search fortinet -vvvAdding repository pypi-dev (http://localhost:8080/simple) and setting it as primary
Deactivating the PyPI repository
Checking if keyring is available
[keyring:keyring.backend] Loading KWallet
[keyring:keyring.backend] Loading SecretService
[keyring:keyring.backend] Loading Windows
[keyring:keyring.backend] Loading chainer
[keyring:keyring.backend] Loading libsecret
[keyring:keyring.backend] Loading macOS
Backend 'fail Keyring' is not suitable
No valid keyring backend was found
Creating new session for localhost:8080
[urllib3:urllib3.connectionpool] Starting new HTTP connection (1): localhost:8080
[urllib3:urllib3.connectionpool] http://localhost:8080 "GET /simple/ HTTP/1.1" 200 318
[urllib3:urllib3.connectionpool] Resetting dropped connection: localhost
[urllib3:urllib3.connectionpool] http://localhost:8080 "GET /simple/fortinet/ HTTP/1.1" 200 629

fortinet (2.1.0.dev3)

```

## Platform: Windows:

I get the below error related to socket permission even from an Administrative terminal

```powershell
(poetry-search) PS C:\Projects\tmp\poetry-private-search> poetry@9949.exe search pytorch
(poetry-search) PS C:\Projects\tmp\poetry-private-search>
(poetry-search) PS C:\Projects\tmp\poetry-private-search>
(poetry-search) PS C:\Projects\tmp\poetry-private-search>
(poetry-search) PS C:\Projects\tmp\poetry-private-search> poetry@9949.exe source add pytorch https://download.pytorch.org/whl/cu113
Adding source with name pytorch.
(poetry-search) PS C:\Projects\tmp\poetry-private-search> poetry@9949.exe search pytorch

HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /whl/cu113/ (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x00000212B238C7A0>: Failed to establish a new connection: [WinError 10013] An attempt was made to access a socket in a way forbidden by its access permissions'))
(poetry-search) PS C:\Projects\tmp\poetry-private-search>
```

**Comment by abn** [comment]
@lumen-sreejithr unfortunately that seems to be more an environment issue since Poetry under the hood simply uses requests to make the request. I do not have a windows environment I can reproduce this on unfortunately. You could try creating a simple script with `requests` to get the `https://download.pytorch.org/whl/cu113/` and run it in the same environment to debug what is going on.

**Comment by lumen-sreejithr** [comment]
@abn , Yes, you were correct. It appears that the issue is unrelated to Poetry and is instead related to my environment. It turns out that there are network restrictions being enforced at my workplace.. So guess its safe to assume same behavior on windows as well.
